h3i/src/actions/h3.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2024, Cloudflare, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+//! Actions specific to HTTP/3 and QUIC
+//!
+//! Actions are small operations such as sending HTTP/3 frames or managing QUIC
+//! streams. Each independent use case for h3i requires its own collection of
+//! Actions, that h3i iterates over in sequence and executes.
+
+use std::collections::HashMap;
+use std::time::Duration;
+
+use qlog::events::EventData;
+use qlog::events::ExData;
+use qlog::events::JsonEvent;
+use quiche;
+use quiche::h3::frame::Frame;
+use quiche::h3::Header;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_with::serde_as;
+
+use crate::encode_header_block;
+
+pub const HTTP3_CONTROL_STREAM_TYPE_ID: u64 = 0x0;
+pub const HTTP3_PUSH_STREAM_TYPE_ID: u64 = 0x1;
+pub const QPACK_ENCODER_STREAM_TYPE_ID: u64 = 0x2;
+pub const QPACK_DECODER_STREAM_TYPE_ID: u64 = 0x3;
+
+pub enum QlogEvent {
```

Comment:
Also, this could use some docs, maybe just a description of what it is.

Reply:
actually these public types are a hangover, so I'll remove them altogether

Reply:
Oh, I found a place we use them internally. We probably want them back

Reply:
We don't need to shadow the record replay type, can you come up with a solution to use just one?

Reply:
Yeah, all good